adam_core/backend/backend.py

```diff
@@ -0,0 +1,168 @@
+import copy
+import logging
+import multiprocessing as mp
+from abc import ABC, abstractmethod
+from typing import Union
+
+from astropy.time import Time
+
+from ..orbits import Orbits
+from ..utils.indexable import concatenate
+from ..utils.multiprocessing import _check_parallel
+
+logger = logging.getLogger(__name__)
+
+
+def propagation_worker(orbits: Orbits, times: Time, backend: "Backend") -> Orbits:
+    propagated = backend._propagate_orbits(orbits, times)
+    return propagated
+
+
+def ephemeris_worker(orbits: Orbits, observers, backend: "Backend"):
+    ephemeris = backend._generate_ephemeris(orbits, observers)
+    return ephemeris
+
+
+class Backend(ABC):
```

Comment:
I think `Backend` is too generic of a name, and risks getting overloaded with tons of stuff. I'd recommend naming this `Propagator`. Ephemeris generation stands out as a little bit extra here (add an `EphemerisGenerator` ABC perhaps?) but I think that's acceptable enough.

Reply:
Renamed to `Propagator`. I'd like to keep ephemeris generation as part of the same class that does propagation because you do both of those together so frequently.